scripts/core/helpers/request-queue.ts

```diff
@@ -0,0 +1,198 @@
+import {httpRequestJsonLocal} from './network';
+
+export interface IRequestParams {
+    method: 'GET' | 'POST';
+    endpoint: string;
+    data?: any;
+    params?: any;
+    id?: string;
+    abortSignal?: AbortSignal;
+}
+
+export enum RequestPriority {
+    LOW,
+    HIGH,
+}
+
+enum RequestStatus {
+    PENDING = 'PENDING',
+    SENDING = 'SENDING',
+    FAILED = 'FAILED',
+    DONE = 'DONE',
+}
+
+class RequestQueue {
+    concurrency = 2;
+    queue: Array<Request> = [];
+    updateTimeout: number | null = null;
+
+    add(params: IRequestParams, provider: any, priority: RequestPriority = RequestPriority.LOW) {
+        const req = new Request(params, provider, priority);
+        const prev = this.queue.find((_req) => _req.isEqual(req) && !_req.isFinished());
+
+        if (prev != null) {
+            return prev.promise;
```

Comment:
Shouldn't existing request be cancelled and a new one initialized in order to avoid old data being returned?

Reply:
that should be before we get the data, but I could check if it's pending and not started yet